sdk/iot/core/tests/cmocka/az_iot_sas_token_tests.c

```diff
@@ -19,7 +22,38 @@
 #define TEST_EXPIRATION_STR "1578941692"
 #define TEST_KEY_NAME "iothubowner"
 
-/*
+#ifndef NO_PRECONDITION_CHECKING
+
+// This block defines the resources needed to verify precondition checking.
+// The macro assert_precondition_checked(func) alone shall be used to assert if a precondition is verified.
+// If a precondition is not verified within a function the result will be either:
+// - A crash in the test, as the the target function will continue to execute with an invalid argument, or
+// - An assert failure indicating a precondition was not tested.
+// Notice that:
+// - If assert_precondition_checked(func) is used, the module must include <setjmp.h>;
+// - If a function has two precondition checks and both are supposed to fail on a given test, assert_precondition_checked(func)
+//   is unable to distinguish which precondition has failed first. Testing precondition checking separately is advised.
+// - Tests using assert_precondition_checked(func) must not be run in parallel (!).
+
+static jmp_buf g_jmp_buf; 
+static unsigned int precondition_test_count = 0;
+static void az_precondition_test_failed_fn()
+{
+  precondition_test_count++;
+  longjmp(g_jmp_buf, 0);
+} 
+
+#define assert_precondition_checked(fn) \
+  precondition_test_count = 0; \
+  (void)setjmp(g_jmp_buf); \
+  if (precondition_test_count == 0) { \
+    fn; \
+  } \
+  assert_int_equal(1, precondition_test_count);
```

Comment:
Because this doesn't return from the function, if two preconditions fail would this evaluate to false? I guess that's a feature not a bug.

Reply:
This mechanism kicks in on the first precondition that fails, no matter which and how many would fail on a given function call.
If two preconditions are to fail, only the first one will actually be checked.